Edit the `trace_signals()` and `report_signals()` in Environment

Krebrov001 opened the pull request:
- Modify these methods to use a common internal `signal_parser_method()`
- Modify the Reporter and Tracer classes where these methods are used.
- Move the common code into the `signal_parser_method()`

- Relates to #2052
- Fixes #2052

Krebrov001 committed:
Edit the `trace_signals()` and `report_signals()` in Environment

- Modify these methods to use a common internal `signal_parser_method()`
- Modify the Reporter and Tracer classes where these methods are used.
- Move the common code into the `signal_parser_method()`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

cmcantalupo commented on src/Environment.hpp:
```diff
@@ -40,8 +42,9 @@ namespace geopm
-            virtual std::string report_signals(void) const = 0;
+            virtual std::vector<std::pair<std::string, int> > trace_signals(void) const = 0;
+            virtual std::vector<std::pair<std::string, int> > report_signals(void) const = 0;
+            virtual std::vector<std::pair<std::string, int> > signal_parser(std::string environment_variable_contents) const = 0;
```
signal_parser belongs in EnvironmentImp, not in the public interface of Environment.

Krebrov001 committed:
Modifying the tests for the updated interfaces.

- move `signal_parser()` from the public interface of `Environment` into `EnvironmentImp`
- modify existing `EnvironmentTests` for the `trace_signals()` and `report_signals()`
- add new test in `EnvironmentTests` for the `signal_parser()` helper method
- modify existing `ReporterTest` to use the updated interface of `report_signals()`
- modify existing `TracerTest` to use the updated interface of `trace_signals()`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

bgeltz committed:
Fix TracerTest

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
Change-Id: I87a0adc71d8b82477c4fcc93b84274d2ec6f103c

bgeltz committed:
Add PlatformIO::signal_exists()

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Add PlatformIO to Environment

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Try to continue if an environment signal is invalid

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
WIP: Do lazy init on Environment::m_platform_io

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

Krebrov001 committed:
Fixed the compilation bugs

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

bgeltz committed:
Make PlatformIO a raw pointer to avoid smart pointer cleanup

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Use MockPlatformIO in EnvironmentTest

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

Krebrov001 committed:
fixed the tests

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

cmcantalupo commented on service/src/geopm/PlatformIO.hpp:
```diff
@@ -248,6 +248,10 @@ namespace geopm
+            /// @brief Query to see if the signal is valid.
+            ///
+            /// @param [in] name Signal to verify.
+            virtual bool signal_exists(const std::string &signal_name) const = 0;
```
Although in principle this interface seems useful, we cannot simply append to the PlatformIO C++ interface.  This interface is exposed in C, Python, and through DBus in addition to the C++ implementation.  To add such a public interface to PlatformIO would require wrappers for all these other APIs.  

In addition, the implementation for this method is easily accomplished with an existing PlatformIO interface in all supported languages with something like:

```
    auto signal_names = platform_io().signal_names();
    bool signal_exists = (signal_names.find(signal_name) != signal_names.end());
```

similar code can be used for controls and the `control_names()` method.  I'm not in favor of adding an interface which contains logic that can be easily derived from another interface.

cmcantalupo commented on src/Environment.cpp:
```diff
@@ -61,12 +65,14 @@ namespace geopm
-    EnvironmentImp::EnvironmentImp(const std::string &default_config_path, const std::string &override_config_path)
+    EnvironmentImp::EnvironmentImp(const std::string &default_config_path,
+                                   const std::string &override_config_path,
+                                   const PlatformIO* platform_io)
```
```suggestion
                                   const PlatformIO *platform_io)
```

cmcantalupo commented on src/Environment.cpp:
```diff
@@ -261,14 +268,47 @@ namespace geopm
+
+        std::vector<std::pair<std::string, int> > result_data_structure;
+
+        auto individual_signals = geopm::string_split(environment_variable_contents, ",");
+        for (const auto& signal : individual_signals) {
+            auto signal_domain = geopm::string_split(signal, "@");
+            if (!m_platform_io->signal_exists(signal_domain[0])) {
```
```suggestion
        auto signals_avail = m_platform_io->signal_names();
        auto individual_signals = geopm::string_split(environment_variable_contents, ",");
        for (const auto &signal : individual_signals) {
            auto signal_domain = geopm::string_split(signal, "@");
            if (signals_avail.find(signal_domain[0]) == signals_avail.end()) {
```

cmcantalupo commented on src/Reporter.cpp:
```diff
@@ -419,25 +419,24 @@ namespace geopm
+                m_env_signal_name_idx.emplace_back(
+                    signal_domain_pair.first,
+                    m_sample_agg->push_signal(signal_domain_pair.first, GEOPM_DOMAIN_BOARD, 0));
+            } else {
```
```suggestion
            }
            else {
```

cmcantalupo reviewed (dismissed):
Please remove signal_exists() from public PlatformIO interface.  Otherwise looks pretty good.

Krebrov001 committed:
Addressing review feedback for Environment edits

- Remove `signal_exists()` from public `PlatformIO` interface.
- Use the `signal_names()` method instead, modifying mocks accordingly.
- `EnvironmentImp::signal_parser()` now throws upon invalid signal name.
- Change the tests to anticipate the thrown exception when testing with invalid signal.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

bgeltz commented on test/EnvironmentTest.cpp:
```diff
@@ -462,8 +573,60 @@ TEST_F(EnvironmentTest, record_filter_off)
+        {"TIME", geopm_domain_e::GEOPM_DOMAIN_BOARD}
+    };
+    environment_variable_contents = "CPU_FREQUENCY_MIN,CPUINFO::FREQ_STEP,TIME";
+    actual_signals = ((EnvironmentImp*)m_env.get())->signal_parser(environment_variable_contents);
```
```suggestion
    actual_signals = m_env->signal_parser(environment_variable_contents);
```

bgeltz commented on test/EnvironmentTest.cpp:
```diff
@@ -462,8 +573,60 @@ TEST_F(EnvironmentTest, record_filter_off)
+        {"TIME", geopm_domain_e::GEOPM_DOMAIN_CORE}
+    };
+    environment_variable_contents = "CPU_FREQUENCY_MIN,CPUINFO::FREQ_STEP@package,TIME@core";
+    actual_signals = ((EnvironmentImp*)m_env.get())->signal_parser(environment_variable_contents);
```
```suggestion
    actual_signals = m_env->signal_parser(environment_variable_contents);
```

bgeltz commented on test/EnvironmentTest.cpp:
```diff
@@ -462,8 +573,60 @@ TEST_F(EnvironmentTest, record_filter_off)
+    };
+    environment_variable_contents = "CPUINFO::FREQ_STEP@invalid,TIME@invalid";
+    GEOPM_EXPECT_THROW_MESSAGE(
+        ((EnvironmentImp*)m_env.get())->signal_parser(environment_variable_contents),
```
```suggestion
        m_env->signal_parser(environment_variable_contents),
```

bgeltz commented on test/EnvironmentTest.cpp:
```diff
@@ -462,8 +573,60 @@ TEST_F(EnvironmentTest, record_filter_off)
+    GEOPM_EXPECT_THROW_MESSAGE(
+        ((EnvironmentImp*)m_env.get())->signal_parser(environment_variable_contents),
+        GEOPM_ERROR_INVALID,
+        (std::string("PlatformTopo::domain_name_to_type(): unrecognized domain_name: ") + std::string("invalid"))
```
```suggestion
        "PlatformTopo::domain_name_to_type(): unrecognized domain_name: invalid")
```

bgeltz commented on test/EnvironmentTest.cpp:
```diff
@@ -462,8 +573,60 @@ TEST_F(EnvironmentTest, record_filter_off)
+
+    environment_variable_contents = "CPU_FREQUENCY_MIN,CPUINFO::FREQ_STEP@package@core,TIME@core";
+    GEOPM_EXPECT_THROW_MESSAGE(
+        ((EnvironmentImp*)m_env.get())->signal_parser(environment_variable_contents),
```
```suggestion
        m_env->signal_parser(environment_variable_contents),
```

bgeltz commented on test/EnvironmentTest.cpp:
```diff
@@ -462,8 +573,60 @@ TEST_F(EnvironmentTest, record_filter_off)
+    GEOPM_EXPECT_THROW_MESSAGE(
+        ((EnvironmentImp*)m_env.get())->signal_parser(environment_variable_contents),
+        GEOPM_ERROR_INVALID,
+        (std::string("EnvironmentImp::signal_parser(): Environment trace extension contains signals with multiple \"@\" characters."))
```
```suggestion
        "EnvironmentImp::signal_parser(): Environment trace extension contains signals with multiple \"@\" characters.")
```

bgeltz commented on test/EnvironmentTest.cpp:
```diff
@@ -462,8 +573,60 @@ TEST_F(EnvironmentTest, record_filter_off)
+
+    environment_variable_contents = "CPU_FREQUENCY_MIN,NUM_VACUUM_TUBES@package,TIME@core";
+    GEOPM_EXPECT_THROW_MESSAGE(
+        ((EnvironmentImp*)m_env.get())->signal_parser(environment_variable_contents),
```
```suggestion
        m_env->signal_parser(environment_variable_contents),
```

bgeltz commented on test/EnvironmentTest.cpp:
```diff
@@ -462,8 +573,60 @@ TEST_F(EnvironmentTest, record_filter_off)
+    GEOPM_EXPECT_THROW_MESSAGE(
+        ((EnvironmentImp*)m_env.get())->signal_parser(environment_variable_contents),
+        GEOPM_ERROR_INVALID,
+        (std::string("Invalid signal : ").append("NUM_VACUUM_TUBES"))
```
```suggestion
        "Invalid signal : NUM_VACUUM_TUBES")
```

Krebrov001 committed:
Changing the strings in `EnvironmentTest.cpp`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Use C++ `dynamic_cast`s instead of C style raw casts

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 commented:
@bgeltz can you please look at this?
